docs: :memo: add docs for data types (#1098)
seedcase-project/seedcase-sprout: docs/design/interface/data-types.qmd

---
title: "Data types"
description: "The data types Sprout supports"
---

> Sprout implements the Frictionless Data Package standard and aims to support the [data types](https://datapackage.org/standard/table-schema/#field-types) it defines. However, Sprout not only describes data with metadata but also transforms it into a tidy Parquet file, ready for querying (see [Outputs](/docs/design/interface/outputs.qmd#files) and [Why Parquet](https://decisions.seedcase-project.org/why-parquet/) for more details). As a result, Sprout supports only data types that are compatible (or can be made compatible) with Parquet storage.

Reviewer: Since Data Package actually calls the “data types” for “field types”, maybe that should be referenced here?

| `time`       | `Time`              | `time64[ns]`                 | `INT64` (Time)         |
| `year`       | `Int32`             | `int32`                      | `INT32`                |
| `yearmonth`  | `Date`              | `date32[day]`                | `INT32` (Date)         |
| `duration`   | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `geopoint`   | `Array[Float64, 2]` | `fixed_size_list[double, 2]` | `FIXED_LEN_BYTE_ARRAY` |
| `geojson`    | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `object`     | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `array`      | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `any`        | `String`            | `string`                     | `BYTE_ARRAY` (String)  |

: Mappings of Frictionless data types in Sprout. In the Parquet column, first the [primitive type](https://parquet.apache.org/docs/file-format/types/) is given, then the [logical type](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md) in brackets, if relevant.

## String

> A sequence of UTF-8 encoded characters. Supported formats: `default`, `email`, `uri`, `binary`, `uuid`.

Reviewer: ```suggestion
A sequence of UTF-8 encoded characters. Sprout supports all Frictionless Data Package string formats: `default`, `email`, `uri`, `binary`, `uuid`.
```
or something like that.

> A sequence of UTF-8 encoded characters. Supported formats: `default`, `email`, `uri`, `binary`, `uuid`.

Reviewer: Add a table with allowed values and examples like for the other data types?

Author: Good idea! Probably best done when I (re)write the format checks, because the examples will depend on what's possible to check :sweat_smile:

| Name                       | Allowed values                          | Examples                   |
|----------------------------|-----------------------------------------|----------------------------|
| decimal indicator          | `.`                                     | `12.56`, `50.000`          |
| leading sign               | `+`(default) or `-`                     | `12.56`, `+5.00`, `-12`    |
| leading and trailing zeros | `0` (optional)                          | `12.56`, `0012`, `12.5600` |
| exponent                   | `E<sign><decimal digits>`               | `5E10`, `12.56E-3`         |
| special values             | `NaN`, `inf`, `-inf` (case-insensitive) |                            |

: `number` format options.

The configuration options `decimalChar`, `groupChar` and `bareNumber` are not yet supported.

## Integer

A whole number with no decimal part.

| Name          | Allowed values | Examples   |
|---------------|----------------|------------|
| leading sign  | no sign or `-` | `5`, `-12` |
| leading zeros | `0` (optional) | `005`      |

: `integer` format options.

> The configuration options `groupChar` and `bareNumber` are not yet supported.

Reviewer: ```suggestion
The Frictionless Data Package configuration options `groupChar` and `bareNumber` are not yet supported by Sprout.
```
Just to make it more explicit? :)

## Boolean

> One of two possible values: true or false. Sprout supports the default notation for truth values in Frictionless:

Reviewer: ```suggestion
One of two possible values: `true` or `false`. Sprout supports the default notation for truth values in Frictionless:
```

Author: I see the idea, but maybe double quotes? Because here I kinda mean the conceptual categories of truthiness and falsy-ness, not any concrete code values.

Reviewer: Ah, I see. Double quotes would be fine for me too 👍

> -   All values in `["true", "True", "TRUE", "1"]` are interpreted as true.
> -   All values in `["false", "False", "FALSE", "0"]` are interpreted as false.

Reviewer: ```suggestion
-   All values in `["true", "True", "TRUE", "1"]` are interpreted as `true`.
-   All values in `["false", "False", "FALSE", "0"]` are interpreted as `false`.
```

> Setting custom `trueValues` and `falseValues` is not yet supported.

Reviewer: ```suggestion
Setting custom `trueValues` and `falseValues` is not yet supported by Sprout.
```

## Datetime

A date with a time and optional timezone.

| Name     | Allowed values                               | Examples                                      |
|----------|----------------------------------------------|-----------------------------------------------|
| without milliseconds or timezone | `YYYY-MM-DDTHH:MM:SS`                | `2002-10-12T12:04:15`, `0202-10-10T02:30:00`                   |
| with milliseconds                | `YYYY-MM-DDTHH:MM:SS.sss`            | `2002-10-12T12:04:15.3`, `0202-10-10T02:30:00.345`             |
| with timezone                    | `YYYY-MM-DDTHH:MM:SS<sign>HH:MM`     | `2002-10-12T12:04:15+05:00`, `0202-10-10T02:30:00-01:00`       |
| with milliseconds and timezone   | `YYYY-MM-DDTHH:MM:SS.sss<sign>HH:MM` | `2002-10-12T12:04:15.3+05:00`, `0202-10-10T02:30:00.345-01:00` |

> | shorthand for UTC                | `YYYY-MM-DDTHH:MM:SS(.sss)Z`         | `2002-10-12T12:04:15Z`, `0202-10-10T02:30:00.345Z`             |

Reviewer: ```suggestion
| with shorthand for UTC                | `YYYY-MM-DDTHH:MM:SS(.sss)Z`         | `2002-10-12T12:04:15Z`, `0202-10-10T02:30:00.345Z`             |
```

: `datetime` format options.

**Restrictions:**

> -   Setting a custom `datetime` pattern in the `format` property is not yet supported. The `any` format is not supported.

Reviewer: ```suggestion
-   Setting a custom `datetime` pattern in the `format` property in `datapackage.json` is not yet supported. The `any` format is not supported.
```
Or?

-   Negative `datetime` values are not supported.
-   Years with more than 4 digits are not supported.

> -   Mixing `datetime` values with and without a timezone in one column is not allowed.

Reviewer: ```suggestion
-   Mixing `datetime` values with and without a timezone in one column are not allowed.
```

Author: Mixing ... is not allowed
Subject is "mixing"

-   When `datetime` values in a column have a timezone, they must all have the same timezone.

## Date

A date without a time.

- Expected format: `YYYY-MM-DD`.

> - Example: `2022-12-09`.

Reviewer: Make this a table like the other data types?

Reviewer: Same for year and yearmonth below

**Restrictions:**

-   Setting a custom `date` pattern in the `format` property is not yet supported. The `any` format is not supported.

> -   Negative `date` values are not supported.

Reviewer 2: Fascinating! I think for now we can say we don't support negative dates... or that we don't guarantee anything :stuck_out_tongue:

-   Years with more than 4 digits are not supported.
-   `date` values with a timezone are not supported.

## Time

A time without a date.

| Name                 | Allowed values    | Examples                        |
|----------------------|-------------------|---------------------------------|
| without microseconds | `HH:MM:SS`        | `12:04:15`                      |
| with microseconds    | `HH:MM:SS.ssssss` | `12:04:15.3`, `02:30:00.345345` |

: `time` format options.

**Restrictions:**

-   Setting a custom `time` pattern in the `format` property is not yet supported. The `any` format is not supported.
-   `time` values with a timezone are not supported.

## Year

A calendar year without month or day.

- Expected format: `YYYY`. Negative `year` values are allowed.
- Examples: `2022`, `-1000`, `0005`.

> 

Reviewer: Add table with examples. Then one of the examples would be a negative year as well.

## Duration

A duration of time.

- Expected pattern: `PnYnMnDTnHnMnS`. See the [definition of the Frictionless type](https://datapackage.org/standard/table-schema/#duration) for more information.
- Example: `P1Y2M3DT10H30M45.343S`.

The number of seconds may include decimal digits to arbitrary precision.

**Restrictions:**

-   The underlying representation of `duration` is `string`. Sprout does not attempt to parse `duration` values into a data type that is aware of the various time units contained within a `duration` value.
-   As a consequence, constraints relying on the numeric comparison of `duration` values are not supported. These constraints are: `minimum`, `maximum`, `exclusiveMinimum`, `exclusiveMaximum`.

::: callout-tip
If you are working with duration or interval values, you could consider converting them to a form that Sprout can parse and compare numerically. Here are some suggestions:

-   If your intervals have start and end points, you could represent them using two `date`, `time` or `datetime` columns. For example, a column called `<column_name>_start` for the beginning of the interval and a column called `<column_name>_end` for the end of the interval.

> -   If it doesn't make sense to represent your duration values as intervals between start and end points, you could represent them as plain `integer`s or `number`s. For example, by calculating the number of days, hours, seconds, milliseconds, etc. that make up your durations.

Reviewer: ```suggestion
-   If it doesn't make sense to represent your duration values as intervals between start and end points, you could represent them as plain `integer`s or `number`s. For example, by calculating the number of days, hours, seconds, milliseconds, etc. (depending on the level of precision you need) that make up your durations.
```

:::

## Geopoint

A geographic point.

- Expected format: `LAT, LONG`. The space is optional.

> - Examples: `45.50, 90.50`, `45.50,90.50`, `-45.50, -90.50`.

Reviewer: Maybe this could be a table as well? :)

Precision of up to 16 significant digits.

| Name                       | Allowed values                          | Examples                   |
|----------------------------|-----------------------------------------|----------------------------|
| decimal indicator          | `.`                                     | `12.56`, `50.000`          |
| leading sign               | `+`(default) or `-`                     | `12.56`, `+5.00`, `-12`    |
| leading and trailing zeros | `0` (optional)                          | `12.56`, `0012`, `12.5600` |
| exponent                   | `E<sign><decimal digits>`               | `5E10`, `12.56E-3`         |
| special values             | `NaN`, `inf`, `-inf` (case-insensitive) |                            |

: `number` format options

The configuration options `decimalChar`, `groupChar` and `bareNumber` are not yet supported.

## Integer

A whole number with no decimal part.

| Name          | Allowed values | Examples   |
|---------------|----------------|------------|
| leading sign  | no sign or `-` | `5`, `-12` |
| leading zeros | `0` (optional) | `005`      |

> : `integer` format options

Reviewer 2: ```suggestion
: `integer` format options.
```

The configuration options `groupChar` and `bareNumber` are not yet supported.

## Boolean

> One of two possible values: true or false.
> 
> Sprout supports the default notation for truth values in Frictionless:

Reviewer 2: ```suggestion
One of two possible values: true or false. Sprout supports the default notation for truth values in Frictionless:
```

-   All values in `["false", "False", "FALSE", "0"]` are interpreted as false.

Setting custom `trueValues` and `falseValues` is not yet supported.

## Datetime

A date with a time and optional timezone.

| Name     | Allowed values                               | Examples                                      |
|----------|----------------------------------------------|-----------------------------------------------|
| without milliseconds or timezone | `YYYY-MM-DDTHH:MM:SS`                | `2002-10-12T12:04:15`, `0202-10-10T02:30:00`                   |
| with milliseconds                | `YYYY-MM-DDTHH:MM:SS.sss`            | `2002-10-12T12:04:15.3`, `0202-10-10T02:30:00.345`             |
| with timezone                    | `YYYY-MM-DDTHH:MM:SS<sign>HH:MM`     | `2002-10-12T12:04:15+05:00`, `0202-10-10T02:30:00-01:00`       |
| with milliseconds and timezone   | `YYYY-MM-DDTHH:MM:SS.sss<sign>HH:MM` | `2002-10-12T12:04:15.3+05:00`, `0202-10-10T02:30:00.345-01:00` |
| shorthand for UTC                | `YYYY-MM-DDTHH:MM:SS(.sss)Z`         | `2002-10-12T12:04:15Z`, `0202-10-10T02:30:00.345Z`             |

> : `datetime` format options

Reviewer 2: ```suggestion
: `datetime` format options.
```

**Restrictions:**

-   Setting a custom `datetime` pattern in the `format` property is not yet supported. The `any` format is not supported.
-   Negative `datetime` values are not supported.
-   Years with more than 4 digits are not supported.
-   Mixing `datetime` values with and without a timezone in one column is not allowed.
-   When `datetime` values in a column have a timezone, they must all have the same timezone.

## Date

A date without a time.

- Expected format: `YYYY-MM-DD`.
- Example: `2022-12-09`.

**Restrictions:**

-   Setting a custom `date` pattern in the `format` property is not yet supported. The `any` format is not supported.
-   Negative `date` values are not supported.
-   Years with more than 4 digits are not supported.
-   `date` values with a timezone are not supported.

## Time

A time without a date.

| Name                 | Allowed values    | Examples                        |
|----------------------|-------------------|---------------------------------|
| without microseconds | `HH:MM:SS`        | `12:04:15`                      |
| with microseconds    | `HH:MM:SS.ssssss` | `12:04:15.3`, `02:30:00.345345` |

> : `time` format options

Reviewer 2: ```suggestion
: `time` format options.
```

-   Negative `yearmonth` values are not supported.
-   Years with more than 4 digits are not supported.
-   `yearmonth` values with a timezone are not supported.

## Duration

A duration of time.

- Expected pattern: `PnYnMnDTnHnMnS`. See the [definition of the Frictionless type](https://datapackage.org/standard/table-schema/#duration) for more information.
- Example: `P1Y2M3DT10H30M45.343S`.

The number of seconds may include decimal digits to arbitrary precision.

**Restrictions:**

-   The underlying representation of `duration` is `string`. Sprout does not attempt to parse `duration` values into a data type that is aware of the various time units contained within a `duration` value.
-   As a consequence, constraints relying on the numeric comparison of `duration` values are not supported. These constraints are: `minimum`, `maximum`, `exclusiveMinimum`, `exclusiveMaximum`.

## Geopoint

A geographic point.

- Expected format: `LAT, LONG`. The space is optional.
- Examples: `45.50, 90.50`, `45.50,90.50`, `-45.50, -90.50`.

The underlying representation is an array of two `number`s.

**Restrictions:**

-   Other `geopoint` formats are not yet supported.

## Array

> A JSON array. Must be well-formed [JSON](http://json.org/).
> 
> The underlying representation is `string`.

Reviewer 2: ```suggestion
A JSON array. Must be well-formed [JSON](http://json.org/). The underlying representation is `string`.
```

## Object

> A JSON object. Must be well-formed [JSON](http://json.org/).
> 
> The underlying representation is `string`.

Reviewer 2: ```suggestion
A JSON object. Must be well-formed [JSON](http://json.org/). The underlying representation is `string`.
```

## Geojson

> A JSON object compliant with the [GeoJSON](http://geojson.org/) or [TopoJSON](https://github.com/topojson/topojson-specification/blob/master/README.md) specification.
> 
> The underlying representation is `string`.

Reviewer 2: ```suggestion
A JSON object compliant with the [GeoJSON](http://geojson.org/) or [TopoJSON](https://github.com/topojson/topojson-specification/blob/master/README.md) specification. The underlying representation is `string`.
```

**Restrictions:**

-   `geojson` values are treated as plain `object`s. They are not checked against the GeoJSON or TopoJSON specification.

## Any

> Unspecified or mixed values.
> 
> The underlying representation is `string`.

Reviewer 2: ```suggestion
Unspecified or mixed values. The underlying representation is `string`.
```

---

> title: "Data types supported by Sprout"

Reviewer 2: ```suggestion
title: "Data types"
description: "The data types Sprout supports"
```

So the sidebar entry item text is smaller.

---

> ### Overview

Reviewer 2: ```suggestion
```

Don't actually need this header.

Below, we list Frictionless data types as used in Sprout and give a precise definition for each. Any differences from the Frictionless specification are noted.

| Frictionless | Polars              | Arrow                        | Parquet                |
|--------------|---------------------|------------------------------|------------------------|
| `string`     | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `number`     | `Float64`           | `float64`                    | `DOUBLE`               |
| `integer`    | `Int64`             | `int64`                      | `INT64`                |
| `boolean`    | `Boolean`           | `bool_`                      | `BOOLEAN`              |

> | `datetime`   | `Datetime`          | `timestamp[ms, tz]`          | `INT64` (Timestamp)    |

Reviewer 2: Parquet's data type surprises me! I'm curious how that gets implemented when loading it. It clearly does, just ... how? haha not important, just curiosity.

Author: I guess whatever loads it has its own mapping? So if it gets a 64-bit integer value that is annotated as a Timestamp, it will interpret it, for example, as the number of milliseconds since UNIX epoch 1970-01-01. Part of "interpreting it like that" can mean being able to format it as a more human-friendly datetime.
At least, this is how I imagine it :sweat_smile:

| `date`       | `Date`              | `date32[day]`                | `INT32` (Date)         |
| `time`       | `Time`              | `time64[ns]`                 | `INT64` (Time)         |
| `year`       | `Int32`             | `int32`                      | `INT32`                |
| `yearmonth`  | `Date`              | `date32[day]`                | `INT32` (Date)         |
| `duration`   | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `geopoint`   | `Array[Float64, 2]` | `fixed_size_list[double, 2]` | `FIXED_LEN_BYTE_ARRAY` |
| `geojson`    | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `object`     | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `array`      | `String`            | `string`                     | `BYTE_ARRAY` (String)  |
| `any`        | `String`            | `string`                     | `BYTE_ARRAY` (String)  |

: Mappings of Frictionless data types in Sprout

> ### String

Reviewer 2: ```suggestion
## String
```

A sequence of UTF-8 encoded characters.

> Supported formats: `default`, `email`, `uri`, `binary`, `uuid`.

Reviewer 2: Is this for the properties?

Author: It's the possible formats for a `string` [here](https://datapackage.org/standard/table-schema/#string). Yeah, you would set it in the `FieldProperties`.
When Sprout supports these, that just means that we check that the given value looks like an email, uuid, etc.

Author: Ah, not exactly. 
Let's say they give resource properties like:
```python
ResourceProperties(
    schema=TableSchemaProperties(
        fields=[
            FieldProperties(name="contact_address", type="string", format="email"),
            FieldProperties(name="first_name", type="string"),
        ]
    )
)
```
Now if they add some data to this resource, values in the column called `contact_address` will be checked for being emails, because that field has `format` set to `email` in the properties. Values in the column `first_name` will not be checked against a specific format, because `format` is not set for this field in the properties.

So whether we do format checks for a column depends on what's in the properties and not on the name of the column or anything in the data itself. And the check is done completely independently of saving the values to Parquet. They will be saved as a string in any case. Without the properties it will not be possible to tell that a particular column should be in email format.

(I'm happy to drop this bit and say we don't support formats for `string` type values if it's too confusing. Especially because these formats are not always easy/possible to check. We can have a regex for checking that an email looks roughly like an email, but the full email regex is so long that we probably don't want to use it. Then, `uri` is such a broad category that even some validation libraries accept any string as an URI...)

> ### Number

Reviewer 2: ```suggestion
## Number
```

A number with or without a decimal part. Precision is not reflected in the number of digits specified in the decimal part (`2.0` is not distinct from `2.00` ).

Precision of up to 16 significant digits.

| Name                       | Allowed values                          | Examples                   |
|----------------------------|-----------------------------------------|----------------------------|
| decimal indicator          | `.`                                     | `12.56`, `50.000`          |
| leading sign               | `+`(default) or `-`                     | `12.56`, `+5.00`, `-12`    |
| leading and trailing zeros | `0` (optional)                          | `12.56`, `0012`, `12.5600` |
| exponent                   | `E<sign><decimal digits>`               | `5E10`, `12.56E-3`         |
| special values             | `NaN`, `inf`, `-inf` (case-insensitive) |                            |

: `number` format options

The configuration options `decimalChar`, `groupChar` and `bareNumber` are not yet supported.

> ### Integer

Reviewer 2: ```suggestion
## Integer
```

A whole number with no decimal part.

| Name          | Allowed values | Examples   |
|---------------|----------------|------------|
| leading sign  | no sign or `-` | `5`, `-12` |
| leading zeros | `0` (optional) | `005`      |

: `integer` format options

The configuration options `groupChar` and `bareNumber` are not yet supported.

> ### Boolean

Reviewer 2: ```suggestion
## Boolean
```

One of two possible values: true or false.

Sprout supports the default notation for truth values in Frictionless:

-   All values in `["true", "True", "TRUE", "1"]` are interpreted as true.
-   All values in `["false", "False", "FALSE", "0"]` are interpreted as false.

Setting custom `trueValues` and `falseValues` is not yet supported.

> ### Datetime

Reviewer 2: ```suggestion
## Datetime
```

|----------|----------------------------------------------|-----------------------------------------------|
| without milliseconds or timezone | `YYYY-MM-DDTHH:MM:SS`                | `2002-10-12T12:04:15`, `0202-10-10T02:30:00`                   |
| with milliseconds                | `YYYY-MM-DDTHH:MM:SS.sss`            | `2002-10-12T12:04:15.3`, `0202-10-10T02:30:00.345`             |
| with timezone                    | `YYYY-MM-DDTHH:MM:SS<sign>HH:MM`     | `2002-10-12T12:04:15+05:00`, `0202-10-10T02:30:00-01:00`       |
| with milliseconds and timezone   | `YYYY-MM-DDTHH:MM:SS.sss<sign>HH:MM` | `2002-10-12T12:04:15.3+05:00`, `0202-10-10T02:30:00.345-01:00` |
| shorthand for UTC                | `YYYY-MM-DDTHH:MM:SS(.sss)Z`         | `2002-10-12T12:04:15Z`, `0202-10-10T02:30:00.345Z`             |

: `datetime` format options

**Restrictions:**

-   Setting a custom `datetime` pattern in the `format` property is not yet supported. The `any` format is not supported.
-   Negative `datetime` values are not supported.
-   Years with more than 4 digits are not supported.
-   Mixing `datetime` values with and without a timezone in one column is not allowed.

> -   When `datetime` values in a column have a timezone, they must all have the same timezone.

Reviewer 2: Oooo, that's interesting! I can see, depending on some very niche cases, where this might be a problem (e.g. collect data on one end of Europe vs the other end, across timezones). For now, not an issue (or we will make sure it's not an issue :stuck_out_tongue: )

Author: Yeah, I was also thinking about this in my [comment](https://github.com/seedcase-project/seedcase-sprout/pull/1098/files#r1974342607) that I forgot to post :sweat_smile:

> -   When `datetime` values in a column have a timezone, they must all have the same timezone.

Reviewer 2: Yea, that sounds very reasonable!

> -   When `datetime` values in a column have a timezone, they must all have the same timezone.

Author: Okay, then we can go with converting to UTC for now. That can be updated to converting to a given timezone pretty easily.

> ### Date

Reviewer 2: ```suggestion
## Date
```

A date without a time.

> Expected format: `YYYY-MM-DD`.
> 
> Example: `2022-12-09`.

Reviewer 2: ```suggestion
- Expected format: `YYYY-MM-DD`.
- Example: `2022-12-09`.
```

Small formatting suggestion.

**Restrictions:**

-   Setting a custom `date` pattern in the `format` property is not yet supported. The `any` format is not supported.
-   Negative `date` values are not supported.
-   Years with more than 4 digits are not supported.
-   `date` values with a timezone are not supported.

> ### Time

Reviewer 2: ```suggestion
## Time
```

A time without a date.

| Name                 | Allowed values    | Examples                        |
|----------------------|-------------------|---------------------------------|
| without microseconds | `HH:MM:SS`        | `12:04:15`                      |
| with microseconds    | `HH:MM:SS.ssssss` | `12:04:15.3`, `02:30:00.345345` |

: `time` format options

**Restrictions:**

-   Setting a custom `time` pattern in the `format` property is not yet supported. The `any` format is not supported.
-   `time` values with a timezone are not supported.

> ### Year

Reviewer 2: ```suggestion
## Year
```

A calendar year without month or day.

> Expected format: `YYYY`. Negative `year` values are allowed.
> 
> Examples: `2022`, `-1000`, `0005`.

Reviewer 2: ```suggestion
- Expected format: `YYYY`. Negative `year` values are allowed.
- Examples: `2022`, `-1000`, `0005`.
```

Formatting suggestion.

**Restrictions:**

-   `year` values with a timezone are not supported.

> ### Yearmonth

Reviewer 2: ```suggestion
## Yearmonth
```

A specific month in a specific year.

> Expected format: `YYYY-MM`.
> 
> Example: `2022-12`.

Reviewer 2: ```suggestion
- Expected format: `YYYY-MM`.
- Example: `2022-12`.
```

> The underlying representation of `yearmonth` is `date`.

Reviewer: Wait, so we won’t store `yearmonth` as `YYYY-MM` but add a day value? 
And how come it’s not sortable? 🤔

> The underlying representation of `yearmonth` is `date`.

Author: We can store it as YYYY-MM but then it has to be a string, because there is no date-based data type for yearmonth in Polars/Parquet. I thiiiink sorting that alphabetically would sort them in the correct chronological order, as long as the number of year digits is fixed and there are no negatives. 

The advantage of storing it as a proper date is that we / whoever analyses the data can perform all date operations on the column.

I don't mind which way!

> The underlying representation of `yearmonth` is `date`.

Reviewer: Ahh, thanks for elaborating! I think it makes (the most) sense to add a date (`-01`) then :)

**Restrictions:**

-   Negative `yearmonth` values are not supported.
-   Years with more than 4 digits are not supported.
-   `yearmonth` values with a timezone are not supported.

> ### Duration

Reviewer 2: ```suggestion
## Duration
```

A duration of time.

> Expected pattern: `PnYnMnDTnHnMnS`. See the [definition of the Frictionless type](https://datapackage.org/standard/table-schema/#duration) for more information.
> 
> Example: `P1Y2M3DT10H30M45.343S`.

Reviewer 2: ```suggestion
- Expected pattern: `PnYnMnDTnHnMnS`. See the [definition of the Frictionless type](https://datapackage.org/standard/table-schema/#duration) for more information.
- Example: `P1Y2M3DT10H30M45.343S`.
```

The number of seconds may include decimal digits to arbitrary precision.

**Restrictions:**

> -   The underlying representation of `duration` is `string`. Sprout does not attempt to parse `duration` values into a data type that is aware of the various time units contained within a `duration` value.

Reviewer 2: I think for now, we just say, we don't support dealing with duration values. And instead suggest making a "start" and "end" columns. Or as you suggest, keep as a string and have it in the documentation that we don't do actual duration data types.

> -   The underlying representation of `duration` is `string`. Sprout does not attempt to parse `duration` values into a data type that is aware of the various time units contained within a `duration` value.

Author: Yeah, I think that's a good idea:
- Allow Frictionless' `duration`, but keep it as a string -- just in case that's useful to someone
- Suggest representing duration either with "start" and "end" columns or with a simple number, depending on the kind of data they have

> -   The underlying representation of `duration` is `string`. Sprout does not attempt to parse `duration` values into a data type that is aware of the various time units contained within a `duration` value.

Author: I added this as a tip here

-   As a consequence, constraints relying on the numeric comparison of `duration` values are not supported. These constraints are: `minimum`, `maximum`, `exclusiveMinimum`, `exclusiveMaximum`.

> ### Geopoint

Reviewer 2: ```suggestion
## Geopoint
```

A geographic point.

> Expected format: `LAT, LONG`. The space is optional.

Author: @lwjohnst86 This is the only thing that needs a decision

> Expected format: `LAT, LONG`. The space is optional.

Author: @lwjohnst86 And this is the last thing to iron out 😊

> Expected format: `LAT, LONG`. The space is optional.
> 
> Examples: `45.50, 90.50`, `45.50,90.50`, `-45.50, -90.50`.

Reviewer 2: ```suggestion
- Expected format: `LAT, LONG`. The space is optional.
- Examples: `45.50, 90.50`, `45.50,90.50`, `-45.50, -90.50`.
```

The underlying representation is an array of two `number`s.

**Restrictions:**

-   Other `geopoint` formats are not yet supported.

> ### Array

Reviewer 2: ```suggestion
## Array
```

A JSON array. Must be well-formed [JSON](http://json.org/).

> The underlying representation is `string`.

Reviewer 2: I think it's totally fine to say we don't do JSON (at least not yet)

> ### Object

Reviewer 2: ```suggestion
## Object
```

A JSON object. Must be well-formed [JSON](http://json.org/).

The underlying representation is `string`.

> ### Geojson

Reviewer 2: ```suggestion
## Geojson
```

A JSON object compliant with the [GeoJSON](http://geojson.org/) or [TopoJSON](https://github.com/topojson/topojson-specification/blob/master/README.md) specification.

The underlying representation is `string`.

**Restrictions:**

-   `geojson` values are treated as plain `object`s. They are not checked against the GeoJSON or TopoJSON specification.

> ### Any

Reviewer 2: ```suggestion
## Any
```

Unspecified or mixed values.

The underlying representation is `string`.

> ### List

Reviewer 2: ```suggestion
## List
```

> Not supported.

Reviewer: Strange!
